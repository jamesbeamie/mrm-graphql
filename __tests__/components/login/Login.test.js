import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import sinon from 'sinon';
import LoginComponent, { Login } from '../../../src/components/login/Login';

describe('Login component', () => {
  const wrapperCode = (
    <MemoryRouter keyLength={0}>
      <LoginComponent />
    </MemoryRouter>
  );
  const wrapper = mount(wrapperCode);

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <MemoryRouter keyLength={0}>
        <LoginComponent />
      </MemoryRouter>,
      div,
    );
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the correct heading text', () => {
    expect(wrapper.find('header>h1')).toHaveLength(1);
    expect(wrapper.find('header>h1').text()).toBe('CONVERGE');
  });

  it('renders the correct mrm introduction text', () => {
    const mrmIntro =
      'Meet the Meeting Room Appthat your meeting room app aspires to be.';
    expect(wrapper.find('#converge-intro')).toHaveLength(1);
    expect(wrapper.find('#converge-intro').text()).toBe(mrmIntro);
  });

  it('renders the tablet image', () => {
    expect(wrapper.find('#dark-tablet img')).toHaveLength(1);
  });

  it('renders login button', () => {
    //   check whether login button exists
    expect(wrapper.find('.btn-signin')).toHaveLength(1);
  });

  it('should call componentWillUnmount when unmounted', () => {
    sinon.spy(Login.prototype, 'componentWillUnmount');
    const loginWrapper = mount(<Login
      location={{
          state: { errorMesage: 'Something Went Wrong' },
          pathname: '/',
        }}
      history={{ push: jest.fn() }}
    />);
    expect(loginWrapper).toMatchSnapshot();
    loginWrapper.unmount();
    expect(Login.prototype.componentWillUnmount).toHaveProperty('callCount', 1);
    Login.prototype.componentWillUnmount.restore();
  });

  it('should display an error when login error occurs and close snackbar on time out', () => {
    const push = jest.fn();
    const loginWrapper = mount(<Login
      location={{
          search: '?error=Something%20Went%20Wrong',
          pathname: '/',
        }}
      history={{ push }}
    />);
    expect(loginWrapper).toMatchSnapshot();
    // find the error message and check whether its the expected error message
    expect(loginWrapper.find('Snackbar').length).toBe(1);
    expect(loginWrapper.find('Snackbar').prop('label')).toBe(loginWrapper.state().loginError);
    // call time out function and check whether snackbar closes
    loginWrapper.find('Snackbar').prop('onTimeout')();
    expect(push).toHaveBeenCalled();
    expect(loginWrapper.state().loginError).toBe(null);
    loginWrapper.update();
    expect(loginWrapper.find('Snackbar')).toHaveLength(0);
    expect(loginWrapper).toMatchSnapshot();
  });
});
