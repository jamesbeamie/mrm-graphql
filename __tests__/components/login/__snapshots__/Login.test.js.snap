// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component renders correctly 1`] = `
<MemoryRouter
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(Login)>
      <Route
        render={[Function]}
      >
        <Login
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "canGo": [Function],
              "createHref": [Function],
              "entries": Array [
                Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
              ],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "index": 0,
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
        >
          <header
            className="home-header"
          >
            <h1>
              CONVERGE
            </h1>
          </header>
          <Component>
            <div
              id="converge-intro"
            >
              Meet the Meeting Room App
              <br />
              that your meeting room app aspires to be.
            </div>
          </Component>
          <LoginButton>
            <a
              href="undefined=undefined"
            >
              <input
                alt="Login With Your Email"
                className="btn-signin"
                src="https://firebasestorage.googleapis.com/v0/b/learning-map-app.appspot.com/o/MrmFrontImages%2FButton.png?alt=media&token=fb7692b6-1465-47c5-94f6-c94aff31c275"
                type="image"
              />
            </a>
          </LoginButton>
          <img
            alt="Dark Tablet"
            id="dark-tablet"
            src="https://firebasestorage.googleapis.com/v0/b/learning-map-app.appspot.com/o/MrmFrontImages%2FDark%20Tablet.png?alt=media&token=75ad56c4-f7f1-4322-91ce-b4850fa0eeb3"
          />
        </Login>
      </Route>
    </withRouter(Login)>
  </Router>
</MemoryRouter>
`;

exports[`Login component should call componentWillUnmount when unmounted 1`] = `
<Login
  history={
    Object {
      "push": [MockFunction],
    }
  }
  location={
    Object {
      "pathname": "/",
      "state": Object {
        "errorMesage": "Something Went Wrong",
      },
    }
  }
>
  <header
    className="home-header"
  >
    <h1>
      CONVERGE
    </h1>
  </header>
  <Component>
    <div
      id="converge-intro"
    >
      Meet the Meeting Room App
      <br />
      that your meeting room app aspires to be.
    </div>
  </Component>
  <LoginButton>
    <a
      href="undefined=undefined"
    >
      <input
        alt="Login With Your Email"
        className="btn-signin"
        src="https://firebasestorage.googleapis.com/v0/b/learning-map-app.appspot.com/o/MrmFrontImages%2FButton.png?alt=media&token=fb7692b6-1465-47c5-94f6-c94aff31c275"
        type="image"
      />
    </a>
  </LoginButton>
  <img
    alt="Dark Tablet"
    id="dark-tablet"
    src="https://firebasestorage.googleapis.com/v0/b/learning-map-app.appspot.com/o/MrmFrontImages%2FDark%20Tablet.png?alt=media&token=75ad56c4-f7f1-4322-91ce-b4850fa0eeb3"
  />
</Login>
`;

exports[`Login component should display an error when login error occurs and close snackbar on time out 1`] = `
<Login
  history={
    Object {
      "push": [MockFunction],
    }
  }
  location={
    Object {
      "pathname": "/",
      "search": "?error=Something%20Went%20Wrong",
    }
  }
>
  <header
    className="home-header"
  >
    <h1>
      CONVERGE
    </h1>
  </header>
  <Component>
    <div
      id="converge-intro"
    >
      Meet the Meeting Room App
      <br />
      that your meeting room app aspires to be.
    </div>
  </Component>
  <LoginButton>
    <a
      href="undefined=undefined"
    >
      <input
        alt="Login With Your Email"
        className="btn-signin"
        src="https://firebasestorage.googleapis.com/v0/b/learning-map-app.appspot.com/o/MrmFrontImages%2FButton.png?alt=media&token=fb7692b6-1465-47c5-94f6-c94aff31c275"
        type="image"
      />
    </a>
  </LoginButton>
  <img
    alt="Dark Tablet"
    id="dark-tablet"
    src="https://firebasestorage.googleapis.com/v0/b/learning-map-app.appspot.com/o/MrmFrontImages%2FDark%20Tablet.png?alt=media&token=75ad56c4-f7f1-4322-91ce-b4850fa0eeb3"
  />
  <ThemedActivableRenderer
    action="Dismiss"
    active={true}
    composeTheme="deeply"
    delay={500}
    label="Please Login With Your Andela Email"
    mapThemrProps={[Function]}
    onClick={[Function]}
    onTimeout={[Function]}
    theme={
      Object {
        "button": "button-error",
        "snackbar": "error",
      }
    }
    timeout={6000}
    type="cancel"
  >
    <ActivableRenderer
      action="Dismiss"
      active={true}
      delay={500}
      label="Please Login With Your Andela Email"
      onClick={[Function]}
      onTimeout={[Function]}
      theme={
        Object {
          "button": "button-error",
          "snackbar": "error",
        }
      }
      timeout={6000}
      type="cancel"
    >
      <Snackbar
        action="Dismiss"
        active={true}
        label="Please Login With Your Andela Email"
        onClick={[Function]}
        onTimeout={[Function]}
        theme={
          Object {
            "button": "button-error",
            "snackbar": "error",
          }
        }
        timeout={6000}
        type="cancel"
      >
        <Portal
          className=""
        />
      </Snackbar>
    </ActivableRenderer>
  </ThemedActivableRenderer>
</Login>
`;

exports[`Login component should display an error when login error occurs and close snackbar on time out 2`] = `
<Login
  history={
    Object {
      "push": [MockFunction] {
        "calls": Array [
          Array [
            "/",
          ],
        ],
      },
    }
  }
  location={
    Object {
      "pathname": "/",
      "search": "?error=Something%20Went%20Wrong",
    }
  }
>
  <header
    className="home-header"
  >
    <h1>
      CONVERGE
    </h1>
  </header>
  <Component>
    <div
      id="converge-intro"
    >
      Meet the Meeting Room App
      <br />
      that your meeting room app aspires to be.
    </div>
  </Component>
  <LoginButton>
    <a
      href="undefined=undefined"
    >
      <input
        alt="Login With Your Email"
        className="btn-signin"
        src="https://firebasestorage.googleapis.com/v0/b/learning-map-app.appspot.com/o/MrmFrontImages%2FButton.png?alt=media&token=fb7692b6-1465-47c5-94f6-c94aff31c275"
        type="image"
      />
    </a>
  </LoginButton>
  <img
    alt="Dark Tablet"
    id="dark-tablet"
    src="https://firebasestorage.googleapis.com/v0/b/learning-map-app.appspot.com/o/MrmFrontImages%2FDark%20Tablet.png?alt=media&token=75ad56c4-f7f1-4322-91ce-b4850fa0eeb3"
  />
</Login>
`;
